picamera2/job.py

```diff
@@ -0,0 +1,37 @@
+from concurrent.futures import Future
+
+
+class Job:
+    def __init__(self, functions, signal_function=None):
+        self._functions = functions
+        self._future = Future()
+        self._future.set_running_or_notify_cancel()
+        if signal_function is not None:
+            self._future.add_done_callback(signal_function)
+
+        # I wonder if there is any useful information we could collect, number
+        # of frames it took for things to finish, maybe intermediate results...
+        self.calls = 0
+
+    def execute(self):
+        """Try to execute this Job. It will return True if it finishes,
+        or False if it needs to be tried again."""
+        assert self._functions, "Job already completed!"
+
+        try:
+            done, result = self._functions[0]()
+            self.calls += 1
+            if done:
+                self._functions.pop(0)
+
+                if not self._functions:
+                    self._future.set_result(result)
+
+        except Exception as e:
```

Comment:
I assume 'Exception' can't be narrowed to a subclass? Since arbitrary functions can be passed?